### PR TITLE
[ADD] udes_stock,udes_stock_refactoring: Migrating refactoring by quantity

### PR DESCRIPTION
**Pull request by armandcela**
Story: 4150

### CODE REVIEW

File: addons/udes_stock/models/stock_move_line.py [hunk 1]
```diff
@@ -589,3 +589,62 @@ def mark_as_done(self, values=None):
...
+        exact_mls = self.filtered(lambda l: l.product_uom_qty == maximum_qty)
+        self -= exact_mls
+        for ml in exact_mls:
+            grouped_mls.append(ml)
+
+        # Next try splitting and grouping to maintain a single ml per group:
+        for ml in self:
+            quantity = ml.product_uom_qty
+            if quantity > maximum_qty:
+                num_full_move_lines = int(quantity // maximum_qty)
```
Comment (kdwyer) [inline addons/udes_stock/models/stock_move_line.py:613]: You could also do 
```python
num_full_move_lines, remainder = divmod(quantity, maximum_qty)
```

Comment (armandcela) [reply addons/udes_stock/models/stock_move_line.py:613]: Agree, its neater as you suggested

**Review by kdwyer — APPROVED** [review]
Couple of suggestions, but LGTM :+1:

File: addons/udes_stock/models/stock_move_line.py [hunk 2]
```diff
@@ -589,3 +589,62 @@ def mark_as_done(self, values=None):
...
+        # Group all (partially) done and cancelled moves first
+        excluded_move_lines = self.filtered(lambda m: m.qty_done or m.state == "done")
+        if excluded_move_lines:
+            grouped_mls.append(excluded_move_lines)
+            self -= excluded_move_lines
+
+        # Sse if any mls are equal to the maximum and add them as individual groups
+        exact_mls = self.filtered(lambda l: l.product_uom_qty == maximum_qty)
+        self -= exact_mls
+        for ml in exact_mls:
```
[the remark this answers is not in the document]
Comment (armandcela) [reply addons/udes_stock/models/stock_move_line.py:606]: grouped_mls is a list [stock_move_line(1), stock_move_line(2)] while exact is recordset stock_move_line(3,4)
At this point i think is better to append , unless we change the type of grouped_mls which impacts some other changes

File: addons/udes_stock/models/stock_move_line.py [hunk 3]
```diff
@@ -589,3 +589,62 @@ def mark_as_done(self, values=None):
             ml.write(ml_vals)
 
         return True
+
+    def _split_and_group_mls_by_quantity(self, maximum_qty):
+        """Split move lines into groups of up to a maximum quantity"""
```
[the remark this answers is not in the document]
Comment (armandcela) [reply addons/udes_stock/models/stock_move_line.py:594]: Don't know when it was first done if there was a specific reason which cannot think any.
At the end is needed only to loop through them, but as far was done as a list didnt see a great improvement to change it as a recordset :)

[the remark this answers is not in the document]
Comment (PeterAlabaster) [reply addons/udes_stock/models/stock_move_line.py]: Ah ok so in 11 it was done this way. No worries

**Review by PeterAlabaster — APPROVED** [review]

File: addons/udes_stock/models/stock_move_line.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kdwyer) [reply addons/udes_stock/models/stock_move_line.py:610]:
```python
grouped_mls.extend(*exact_mls)
```
might work, but whatever is easiest.

File: addons/udes_stock/models/stock_move_line.py [hunk 4]
```diff
@@ -589,3 +589,65 @@ def mark_as_done(self, values=None):
...
+        """
+        grouped_mls = []
+
+        # Group all (partially) done and cancelled moves first
+        excluded_move_lines = self.filtered(lambda m: m.qty_done or m.state == "done")
+        if excluded_move_lines:
+            grouped_mls.append(excluded_move_lines)
+            self -= excluded_move_lines
+
+        # Sse if any mls are equal to the maximum and add them as individual groups
```
Comment (ws-mps) [inline addons/udes_stock/models/stock_move_line.py:607]: Typo: Sse

**Comment by ws-mps** [comment]
LGTM using move_lines_for_qty() looks a lot neater.

**Review by ws-mps — APPROVED** [review]

**Pull request merged**